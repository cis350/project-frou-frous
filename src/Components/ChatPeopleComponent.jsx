import './chat.css';
import React, { useState, useEffect } from 'react';
import { getChatFriends } from '../api/chatAPI';

function convertTimestamp(timestamp) {
  console.log('timestamp', timestamp);
  if (timestamp < 0) {
    return '';
  }
  const d = new Date(timestamp);
  console.log('timestamp date', d);
  const yyyy = d.getFullYear();
  const mm = ('0'.concat((d.getMonth() + 1))).slice(-2);
  const dd = ('0'.concat(d.getDate())).slice(-2);

  // ie: 2014-03-24, 3:00 PM
  // ['-', mm.toString(), '-', dd.toString(), ', ', h.toString(), ':', min.toString(), ' ', ampm]
  const time = yyyy.toString().concat('-', mm.toString(), '-', dd.toString());
  return time;
}

export default function ChatPeopleComponent() {
  const curUser = sessionStorage.getItem('username');
  // add a state to the component
  const [search, setSearch] = useState('');
  const [friends, setFriends] = useState([]);

  let people = '';

  function filterFriends(searchString, res) {
    let response = [];
    if (searchString.trim() !== '') {
      response = res.filter((x) => x.friend.startsWith(searchString));
    } else {
      response = res;
    }
    for (let i = 0; i < response.length; i += 1) {
      // people.push(<h2>{response.friends[i].message}</h2>);
      people = people.concat(
        `<a class='chatA' href = '/chat/user/${response[i].friend}'>
          <div class="person">
            <img alt="friendimg" class="circle-pic" src=${response[i].friendImage} />
            <div class="personInfo">
              <div class="nameTime">
                <h3>${response[i].friend}</h3>
                <p>${convertTimestamp(response[i].timestamp)}</p>
              </div>
              <p class="lastMessage">${response[i].message}</p>
            </div>
          </div>
        </a>`,
      );
    }
    document.getElementById('peopleContainers').innerHTML = people;
    people = '';
  }

  function searchFriends(searchString) {
    setSearch(searchString);
    filterFriends(searchString, friends);
  }

  useEffect(() => {
    // Fetches data from the openweathermap API and updates the DOM
    console.log('USING EFFECT');

    const getFriends = async () => {
      try {
        const res = await getChatFriends(curUser);
        setFriends(res.friends);
        // setFriends(response.friends);
        filterFriends(search, res.friends);
      } catch (err) {
        console.error('error', err.message);
      }
    };

    getFriends();
  }, []);
  console.log('PEOPLE', people);
  return (
    <div id="peopleList">
      <div id="searchBarWrapper">
        <div id="searchBar" className="input-group rounded-8">
          <div id="searchIcon">
            <img alt="search" src="https://th.bing.com/th/id/R.602ef64bc31a62f9ebd523d97fc9f369?rik=JAPpFrIn7Yon0Q&riu=http%3a%2f%2fcdn.onlinewebfonts.com%2fsvg%2fimg_194915.png&ehk=PY%2fOHmWzAr5CQnqxsnvx5nnpZAeNl7OJ5%2fHDzvIuHTo%3d&risl=&pid=ImgRaw&r=0" />
          </div>
          <input id="chatSearch" value={search} onChange={(e) => searchFriends(e.target.value)} type="text" className="border-0" style={{ textAlign: 'right' }} placeholder="People Search" />
        </div>
      </div>
      <br />

      <div id="peopleContainers" />
    </div>
  );
}
