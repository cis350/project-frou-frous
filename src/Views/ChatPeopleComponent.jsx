import './chat.css';
import React, { useState, useEffect } from 'react';
import { getChatFriends } from '../api/chatAPI';

function convertTimestamp(timestamp) {
  const d = new Date(timestamp * 1000);
  const yyyy = d.getFullYear();
  const mm = ('0'.concat((d.getMonth() + 1))).slice(-2);
  const dd = ('0'.concat(d.getDate())).slice(-2);

  // ie: 2014-03-24, 3:00 PM
  // ['-', mm.toString(), '-', dd.toString(), ', ', h.toString(), ':', min.toString(), ' ', ampm]
  const time = yyyy.toString().concat('-', mm.toString(), '-', dd.toString());
  return time;
}

export default function ChatPeopleComponent() {
  // add a state to the component
  const [search, setSearch] = useState('');
  const [friends, setFriends] = useState([]);

  let people = '';

  function filterFriends(searchString, res) {
    let response = [];
    if (searchString.trim() !== '') {
      response = res.filter((x) => x.friend.startsWith(searchString));
    } else {
      response = res;
    }
    for (let i = 0; i < response.length; i += 1) {
      // people.push(<h2>{response.friends[i].message}</h2>);
      people = people.concat(
        `<a class='chatA' href = '/chat/user/${response[i].friend}'>
          <div class="person">
            <img alt="friendimg" class="circle-pic" src=${response[i].friendImage} />
            <div class="personInfo">
              <div class="nameTime">
                <h3>${response[i].friend}</h3>
                <p>${convertTimestamp(response[i].timestamp)}</p>
              </div>
              <p class="lastMessage">${response[i].message}</p>
            </div>
          </div>
        </a>`,
      );
    }
    document.getElementById('peopleContainers').innerHTML = people;
    people = '';
  }

  function searchFriends(searchString) {
    setSearch(searchString);
    filterFriends(searchString, friends);
  }

  useEffect(() => {
    // Fetches data from the openweathermap API and updates the DOM
    console.log('USING EFFECT');

    const getFriends = async () => {
      try {
        const res = await getChatFriends();
        setFriends(res.friends);
        // setFriends(response.friends);
        filterFriends(search, res.friends);
      } catch (err) {
        console.error('error', err.message);
      }
    };

    getFriends();
  }, []);
  console.log('PEOPLE', people);
  return (
    <div id="peopleList">
      <div id="searchBarWrapper">
        <div id="searchBar" className="input-group rounded-8">
          <div id="searchIcon">
            <p style={{ color: 'black' }} className="bi bi-search" />
          </div>
          <input id="chatSearch" value={search} onChange={(e) => searchFriends(e.target.value)} type="text" className="border-0" style={{ textAlign: 'right' }} placeholder="People Search" />
        </div>
      </div>
      <br />

      <div id="peopleContainers" />
    </div>
  );
}
