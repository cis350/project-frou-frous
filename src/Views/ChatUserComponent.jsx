import './chat.css';
import React, { useEffect } from 'react';
import { useParams } from 'react-router-dom';
import { getChatMessages, sendChatMessage } from '../api/userPageAPI';
import ChatPeopleComponent from './ChatPeopleComponent';

export default function ChatUserComponent() {
  const user = useParams();

  function scrollMessages() {
    const messageDiv = document.getElementById('messages');
    const sheight = messageDiv.scrollHeight;
    messageDiv.scrollTop = sheight;
  }

  const getChats = async () => {
    try {
      let chats = '';
      const res = await getChatMessages(user.user);
      for (let i = 0; i < res.messages.length; i += 1) {
        let leftRight = 'messageLeft';
        const message = res.messages[i];
        if (message.sent) {
          leftRight = 'messageRight';
        }
        chats = chats.concat(`
          <div class='${leftRight} message'>
              <p> ${message.message} </p>
          </div>
          `);
      }
      document.getElementById('messages').innerHTML = chats;
      scrollMessages();
    } catch (err) {
      console.log(err.message);
    }
  };

  useEffect(() => {
    getChats();
  }, 5000);

  async function sendMessage() {
    const chat = document.getElementById('chatSend');
    if (chat) {
      if (chat.value !== '') {
        try {
          await sendChatMessage(chat.value);
          await getChats();
          chat.value = '';
          chat.focus();
        } catch (err) {
          alert(err.message);
        }
      }
    }
  }

  return (
    <div id="chatWrapper">
      <div id="peopleContainer" style={{ width: '30%' }}>
        <ChatPeopleComponent />
      </div>

      <div id="chatWindow">
        <div id="chatBar" className="flex-row">
          <h3>{user.user}</h3>
          <a className="chatA" href="/user/1234">
            <p id="homeIcon" className="bi bi-house-fill" />
          </a>
        </div>

        <div id="messages" />

        <div id="sendWrapper">
          <div id="sendMessageBar" className="input-group rounded-8">
            <input type="text" id="chatSend" className="border-0" style={{ textAlign: 'left' }} />
            <button type="submit" onClick={sendMessage} onSubmit={sendMessage}>
              <svg xmlns="http://www.w3.org/2000/svg" width="16" height="16" fill="black" className="bi bi-send" viewBox="0 0 16 16">
                <path d="M15.854.146a.5.5 0 0 1 .11.54l-5.819 14.547a.75.75 0 0 1-1.329.124l-3.178-4.995L.643 7.184a.75.75 0 0 1 .124-1.33L15.314.037a.5.5 0 0 1 .54.11ZM6.636 10.07l2.761 4.338L14.13 2.576 6.636 10.07Zm6.787-8.201L1.591 6.602l4.339 2.76 7.494-7.493Z" />
              </svg>
            </button>
          </div>
        </div>

      </div>

    </div>

  );
}
