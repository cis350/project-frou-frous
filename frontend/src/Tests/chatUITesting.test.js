/**
* @jest-environment jsdom
*/

import React from 'react';
import '@testing-library/jest-dom/extend-expect';
// import testing library functions
import { render } from '@testing-library/react';
import renderer, { act } from 'react-test-renderer';

import ChatPeopleComponent from '../Components/ChatPeopleComponent';
import ChatUserComponent from '../Components/ChatUserComponent';

test('test that button present', async () => {
  // render the component
  const { getByRole } = await act(async () => render(<ChatUserComponent />));
  // find the element by its role
  const inputBox = getByRole('button');
  // assert that the element is in the document
  expect(inputBox).toBeInTheDocument();
});

test('snapshot test people', () => {
  const component = renderer.create(<ChatPeopleComponent />);
  const domTreeJSON = component.toJSON();
  // matcher
  expect(domTreeJSON).toMatchSnapshot();
});
