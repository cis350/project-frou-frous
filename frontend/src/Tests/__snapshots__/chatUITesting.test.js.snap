// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test people 1`] = `
<div
  id="peopleList"
>
  <div
    id="searchBarWrapper"
  >
    <div
      className="input-group rounded-8"
      id="searchBar"
    >
      <div
        id="searchIcon"
      >
        <img
          alt="search"
          src="https://th.bing.com/th/id/R.602ef64bc31a62f9ebd523d97fc9f369?rik=JAPpFrIn7Yon0Q&riu=http%3a%2f%2fcdn.onlinewebfonts.com%2fsvg%2fimg_194915.png&ehk=PY%2fOHmWzAr5CQnqxsnvx5nnpZAeNl7OJ5%2fHDzvIuHTo%3d&risl=&pid=ImgRaw&r=0"
        />
      </div>
      <input
        className="border-0"
        id="chatSearch"
        onChange={[Function]}
        placeholder="People Search"
        style={
          {
            "textAlign": "right",
          }
        }
        type="text"
        value=""
      />
    </div>
  </div>
  <br />
  <div
    id="peopleContainers"
  />
</div>
`;
